### ISSUE TITLE
spinlock deadlock with pthread_cancel

### ISSUE THREAD

**Issue by chabbimilind** [HPCToolkit/hpctoolkit#106]
If pthread_cancel is issued to a thread holding the file lock, the thread aborts without releasing the spinlock.
Subsequent threads can never acquire the file i/o lock. 
It is easy to reproduce this with the "-t" flag to create trace files and sending pthread_cancel to the thread during its start-up sequence. 
I encountered it in a production code that uses side threads for I/O and sends pthread_cancel once the main thread comes out of poll().

The fix is to use pthread_cleanup_push/pop (http://man7.org/linux/man-pages/man3/pthread_cleanup_push.3.html) similar to what is done in libmonitor: https://github.com/HPCToolkit/libmonitor/blob/d28cc1d3c08c02013a68a022a57a6ac73db88166/src/pthread.c#L968

File I/O lock is perhaps just one place where it happens but I suspect there could be other places where this issues can happen.

How to reproduce:

run: hpcrun -e REALTIME@5000 -t ./example 100

example.cpp contents:

#include<pthread.h>
#include <iostream>
#include <stdlib.h>
#include <stropts.h>
#include <poll.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

void * dummy1(void *){
        for(int i = 0; i < 1000000; i++);
        return 0;
}

int main(int argc, char **argv){
        long n = 100; //atol(argv[1]);
        struct pollfd d;
        d.fd = open("/dev/null", O_RDWR);
        d.events = POLLIN;
        for(long i = 0; i < n; i++){
                pthread_t x;
                pthread_create(&x, 0, dummy1, (void*)0);
                poll(&d,1, -1);
                pthread_cancel(x);
                pthread_join(x, 0);

        }
        return 0;
}

How to compile:
g++ example.cpp  -o example -lpthread

Backtrace of the deadlocked thread:
(gdb) bt
#0  0x00007f405eb29e5d in spinlock_lock (l=0x7f405ed706c8 <files_lock>) at /home/mc29/software/HPCToolkit/hpctoolkit/BUILD/../src/lib/prof-lean/spinlock.h:115
#1  0x00007f405eb2a763 in hpcrun_open_trace_file (thread=2) at ../../../../src/tool/hpcrun/files.c:455
#2  0x00007f405eb33eca in hpcrun_trace_open (cptd=0x7f405ef80098) at ../../../../src/tool/hpcrun/trace.c:142
#3  0x00007f405eb33326 in hpcrun_thread_data_init (id=2, thr_ctxt=0x7f405c677f40, is_child=0, n_sources=1) at ../../../../src/tool/hpcrun/thread_data.c:371
#4  0x00007f405eb337dc in allocate_thread_data (id=2, thr_ctxt=0x7f405c677f40, num_sources=1) at ../../../../src/tool/hpcrun/threadmgr.c:144
#5  0x00007f405eb33a06 in hpcrun_threadMgr_data_get (id=2, thr_ctxt=0x7f405c677f40, num_sources=1) at ../../../../src/tool/hpcrun/threadmgr.c:264
#6  0x00007f405eb27dee in hpcrun_thread_init (id=2, local_thread_data=0x7f405c677f50) at ../../../../src/tool/hpcrun/main.c:675
#7  0x00007f405eb28aca in monitor_init_thread (tid=2, data=0x7f405c677f50) at ../../../../src/tool/hpcrun/main.c:1104
#8  0x00007f405e8fbd61 in monitor_begin_thread (arg=0x7f405eb09310 <monitor_init_tn_array+80>) at ../../libmonitor/src/pthread.c:974
#9  0x00007f405e6d6184 in start_thread (arg=0x7f405be98700) at pthread_create.c:312
#10 0x00007f405e0efffd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111

**Comment by jmellorcrummey** [comment]
What file lock?

**Comment by chabbimilind** [comment]
Trace file.

**Comment by jmellorcrummey** [comment]
I looked briefly at this problem. Handling pthread cancellation requires more than just push/pop of cancellation handlers for hpctoolkit. 

First, the lock that you mention above has no concept of an owner. In a cancellation handler, a thread would have no idea if it acquired the lock or not. There is no way for the thread to set a flag before or after acquiring the lock. The thread could be cancelled between the two operations and end up doing the wrong thing. The lock could be replaced with one that would acquire ownership by using a CAS to put a thread id into the lock to own it. We would have to ensure that thread ids start at 1 rather than 0, which I think is the case now.

Cancellation is also used libmonitor to signal threads to record record finish their monitoring. We have encountered problems with the signal from libmonitor being delivered inside a signal handler while the thread is taking a sample.

I think the best approach to address both of these issues is to disable cancellation while hpctoolkit is doing its work. For the problems reported above where cancellation arrives early while a thread is holding a lock while manipulating files, disabling cancellation in libmonitor prior to calling monitor_init_thread and then reenabling it afterward would do the trick. Or, cancellation could be disabled/enabled in hpcrun_thread_init. The reason it seems attractive to handle this in libmonitor is that for handling signal delivery for samples, rather than having every sample source enable/disable cancellation, it could be done once for all appropriate signals in libmonitor.

@mwkrentel and I should discuss how to address this issue.
